sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala

```diff
@@ -178,6 +257,47 @@ case class FlatMapGroupsWithStateExec(
       }
     }
 
+    /**
+     * Process the new data iterator along with the initial state. The initial state is applied
+     * before processing the new data for every key. The user defined function is called only
+     * once on the data.
+     */
+    def processNewDataWithInitState(
+        childDataIter: Iterator[InternalRow],
+        initStateIter: Iterator[InternalRow]
+      ): Iterator[InternalRow] = {
+
+      if (!childDataIter.hasNext && !initStateIter.hasNext) return Iterator.empty
+
+      val groupedChildDataIter = GroupedIterator(childDataIter, groupingAttributes, child.output)
+      val groupedInitStateIter =
+        GroupedIterator(initStateIter, initStateGroupAttrs, initialState.output)
+
+      val keyOrderingComparator = GenerateOrdering.generate(
+        groupingAttributes.map(SortOrder(_, Ascending)), groupingAttributes)
+
+      FlatMapGroupsWithStateExec.mergeGroupedIters(
+          groupedChildDataIter,
+          groupedInitStateIter,
+          keyOrderingComparator).flatMap { case (keyRow, valueRowIter, initStateRowOption) =>
+        val keyUnsafeRow = keyRow.asInstanceOf[UnsafeRow]
+        var foundInitStateForKey = false
+        initStateRowOption.foreach { initStateRow =>
+          if (foundInitStateForKey) {
+            throw new IllegalArgumentException("The initial state provided contained duplicates")
```

Comment:
better error message. .. what does duplicates mean. be more verbose... contains multiple state values for a key.